### PR TITLE
fix(a11y): focus monitor not working inside shadow dom

### LINKED ISSUE #19133: Angular Material focus styles not applying in Shadow DOM
# 🐞 bug report

### Affected Package

 This issue is caused by angular/components/pull/18667

### Is this a regression?

 Yes.

### Description

 Keyboard focus and click focus styles no longer apply to Angular Material components when using Shadow DOM. Possibly due to the change in focus using walk down the DOM.

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/no-focus-style-for-interactive-element-shadowdom

### PR DESCRIPTION
**Pull request by crisbeto**
In #18667 event delegation was implemented for the `FocusMonitor` which is based around binding a single event on the `document`. The problem is that the browser won't invoke the `focus` and `blur` handlers on the `document`, if focus is moved within the same shadow root. These changes switch to delegating the events either through the closest shadow root or the `document`.

Fixes #19133.

**Note:** I spent some time trying to reproduce this in a unit test, but I wasn't able to because the way we simulate focus programmatically isn't how things behave in the browser, and we have to use programmatic focus in order to avoid flakes on the CI.

### CODE REVIEW

File: src/cdk/a11y/focus-monitor/focus-monitor.ts [hunk 1]
```diff
@@ -110,6 +111,14 @@ export class FocusMonitor implements OnDestroy {
   /** The number of elements currently being monitored. */
   private _monitoredElementCount = 0;
 
+  /**
+   * Keeps track of the root nodes to which we've currently bound a focus/blur handler,
+   * as well as the number of monitored elements that they contain. We have to treat focus/blur
+   * handlers differently from the rest of the events, because the browser won't emit events
+   * to the document when focus moves inside of a shadow root.
+   */
+  private _rootNodeFocusListenerCount = new Map<HTMLElement|Document, number>();
```
Comment (mmalerba) [inline src/cdk/a11y/focus-monitor/focus-monitor.ts:120]: To be safe, we should scan through this in `ngOnDestroy` and unlisten to any remaining entries

File: src/cdk/a11y/focus-monitor/focus-monitor.ts [hunk 2]
```diff
@@ -225,20 +231,26 @@ export class FocusMonitor implements OnDestroy {
...
+    // to the `document`, if focus is moving within the same shadow root.
+    const rootNode = (_getShadowRoot(nativeElement) as HTMLElement|null) || this._getDocument();
+
     // Check if we're already monitoring this element.
     if (this._elementInfo.has(nativeElement)) {
-      const cachedInfo = this._elementInfo.get(nativeElement);
-      cachedInfo!.checkChildren = checkChildren;
-      return cachedInfo!.subject.asObservable();
+      const cachedInfo = this._elementInfo.get(nativeElement)!;
+      cachedInfo.checkChildren = checkChildren;
```
Comment (mmalerba) [inline src/cdk/a11y/focus-monitor/focus-monitor.ts:242]: I see that you didn't change this, but is this right? Won't this cause all previous calls to monitor on this element to behave according to the `checkChildren` parameter on the current call?

File: src/cdk/a11y/focus-monitor/focus-monitor.ts (hunk 1 above)
Comment (crisbeto) [reply src/cdk/a11y/focus-monitor/focus-monitor.ts:120]: We're looping through and calling `stopMonitoring` on all the remaining elements which eventually empties this out as well.

File: src/cdk/a11y/focus-monitor/focus-monitor.ts (hunk 2 above)
Comment (crisbeto) [reply src/cdk/a11y/focus-monitor/focus-monitor.ts:242]: Judging by how it's written, I think that's the point. E.g. you could turn a shallow watch into a deep one by calling `monitor` again with `checkChildren`. We could separate it out and have 2 separate streams, but we'd have to make a cache key out of the options.

Comment (mmalerba) [reply src/cdk/a11y/focus-monitor/focus-monitor.ts:242]: Yeah, I'm not sure it should work that way. The main way I see it being called twice is if there's e.g. a directive and a component that need to know about focus state on the same element. They might have different requirements about whether to include the children, and we wouldn't want them to affect each other's subscriptions.

So it seems like a bug, but a pre-existing one and I haven't seen anyone complain about it yet, so I think its safe to leave this for another PR.

**Pull request merged**